Add dedicated chapter on directory layout

marcphilipp opened the pull request:
This PR adds a dedicated chapter on directory layout that describe the rough structure of the Gradle user home and project root directories and then documents the new cache cleanup strategies.

oehme commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -49,12 +49,7 @@ We are now confident that the incremental compiler is ready to be used in every
-- Similarly, after building a project, version-specific cache directories in `PROJECT_DIR/.gradle/<gradle-version>/` are checked periodically (at most every 24 hours) for whether they are still in use. They are deleted if they haven't been used for 7 days.
-- Shared versioned cache directories in `GRADLE_USER_HOME/caches/` (e.g. `jars-*`) are checked periodically (at most every 24 hours) for whether they are still in use. If there's no Gradle version that still uses them, they are deleted.
-- Files in shared caches used by the current Gradle version in `GRADLE_USER_HOME/caches/` (e.g. `jars-3` or `modules-2`) are checked periodically (at most every 24 hours) for when they were last accessed. Depending on whether the file can be recreated locally or would have to be downloaded from a remote repository again, it will be deleted after 7 or 30 days of not being accessed, respectively.
+Caching has always been one of the strong suits of Gradle. Over time, more and more persistent caches have been added to improve performance and support new features, requiring more and more disk space on build servers and developer workstations. Gradle now addresses one of the most highly voted issues on GitHub and introduces the cleanup strategies for the caches in the [Gradle user home directo ...
```
Not a blocker, but I'd remove the introductory sentence and start with the headline - We clean up caches now.

eriwen commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+
+The following cleanup strategies are applied periodically (at most every 24 hours):
+
+- Version-specific caches in `caches/<gradle-version>/` are checked for whether they are still in use. If not, directories for release versions are deleted after 30 days of inactivity, snapshot versions after 7 days of inactivity.
```
:o: is there a brief way to define "still in use"?

eriwen commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+
+The project root directory has roughly the following structure:
+
+[listing,subs=+macros]
```
❓ I'm not familiar with this. What's "subs=+macros" for?

pledbrook replied:
@eriwen Enables macro substitution in source code, specifically for `pass:quotes[_or_]` in this case.

eriwen replied:
huh.. the more you know ☄️ 🌟

eriwen commented on subprojects/docs/src/main/resources/header.html:
```diff
@@ -113,6 +113,7 @@ <h3 id="using-gradle-builds">Running Gradle Builds</h3>
             <li><a href="https://docs.gradle.com/build-scan-plugin">Using Build Scans</a></li>
             <li><a href="/userguide/build_cache.html">Enabling and Configuring the Build Cache</a></li>
             <li><a href="/userguide/composite_builds.html">Integrating Separate Gradle Builds (Composite Builds)</a></li>
+            <li><a href="/userguide/directory_layout.html">Directory Layout</a></li>
```
This deserves a more prominent place IMO. Definitely above "Executing Multi-Project Builds" and maybe should be part of "Getting Started". Do you agree @pledbrook?

pledbrook replied:
I think this section of the nav is already overcrowded, and as I mention above, I don't see the value of having this so prominent. Also, I don't think "Directory Layout" is a sufficient chapter title as it could easily be interpreted — and more likely to be interpreted in my view — as about a _build_'s or _project_'s directory layout.

As I mention on [issue #262](https://github.com/gradle/dotorg-docs/issues/262), I think details of the internals of Gradle, such as the wrapper, daemon, and the working directories it uses should be aggregated into an appendix, or perhaps their own reference section.

marcphilipp replied:
@pledbrook Do you have a proposal for a better chapter title?

pledbrook replied:
No really good ones. "Gradle's hidden directories/files"? "Gradle's directories and files"? "Gradle's caches and other working files"? "The directories Gradle uses for internal purposes"?!

I don't know if there is a good, single word that describes what these directories and files are. I'm sure there must be a technical term for such files that's well known, but nothing's popping into my head.

eriwen commented on subprojects/docs/src/docs/userguide/userguide_single.adoc:
```diff
@@ -50,6 +50,8 @@ include::gradle_wrapper.adoc[leveloffset=+2]
 
 include::troubleshooting.adoc[leveloffset=+2]
 
+include::directory_layout.adoc[leveloffset=+2]
```
:o: Here I'd say this probably belongs just under build_environment.

pledbrook replied:
Do users really need to know this information to make good use of Gradle? I get by without most of it. I used to check the modules cache once upon a time, but I'm not sure even that is particularly useful now.

I would much prefer to see this treated as more of an appendix: a reference for those who may find this information interesting or helpful, but not something many users need to worry about.

eriwen commented:
BTW, this is a very useful chapter. I am ashamed we haven't had something like this before.

oehme commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -49,12 +49,7 @@ We are now confident that the incremental compiler is ready to be used in every
-- Similarly, after building a project, version-specific cache directories in `PROJECT_DIR/.gradle/<gradle-version>/` are checked periodically (at most every 24 hours) for whether they are still in use. They are deleted if they haven't been used for 7 days.
-- Shared versioned cache directories in `GRADLE_USER_HOME/caches/` (e.g. `jars-*`) are checked periodically (at most every 24 hours) for whether they are still in use. If there's no Gradle version that still uses them, they are deleted.
-- Files in shared caches used by the current Gradle version in `GRADLE_USER_HOME/caches/` (e.g. `jars-3` or `modules-2`) are checked periodically (at most every 24 hours) for when they were last accessed. Depending on whether the file can be recreated locally or would have to be downloaded from a remote repository again, it will be deleted after 7 or 30 days of not being accessed, respectively.
+Caching has always been one of the strong suits of Gradle. Over time, more and more persistent caches have been added to improve performance and support new features, requiring more and more disk space on build servers and developer workstations. Gradle now addresses one of the most highly voted issues on GitHub and introduces the cleanup strategies for the caches in the [Gradle user home directo ...
```
These are the release notes.

eriwen replied:
Right. Carry on then, nothing to see here 🙈

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+[[directory_layout]]
+= Directory Layout
+
+[.lead]
```
I believe this is unnecessary. The first paragraph after the title is automatically treated as lead style.  Not that I'm a fan of lead style anyway, at least not for the user manual.

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+// limitations under the License.
+
+[[directory_layout]]
+= Directory Layout
```
I would add a preamble about what the aim of this chapter is. What directory layout are we talking about? That's what a reader wants to know.

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+= Directory Layout
+
+[.lead]
+Gradle uses two main directories to perform and manage its work: the <<#directory_layout:gradle_user_home_directory>> and the <<#directory_layout:project_root_directory>>.
```
This sounds like it's supposed to lead into the sections, but doesn't quite do so. I  would follow this with something like

> We look at these two in detail in the following sections.

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+Gradle uses two main directories to perform and manage its work: the <<#directory_layout:gradle_user_home_directory>> and the <<#directory_layout:project_root_directory>>.
+
+
+[[directory_layout:gradle_user_home_directory]]
```
Links from other chapters need to include the name of the chapter file anyway, so the `directory_layout:` prefix is kind of redundant. It does help keep the anchor unique for the single-page HTML and PDF documents, but perhaps `dir:` would be sufficient?

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+<7> Global <<build_environment.adoc#sec:gradle_configuration_properties,Gradle configuration properties>>
+
+[[directory_layout:gradle_user_home_directory:cache_cleanup]]
+=== Global cache cleanup
```
This transition is a little awkward. It feels like the document should be discussing some of the detail of the different directories. I would rather remove the "info" parts of the above annotations and insert a definition list after the directory structure that explains what each directory does and perhaps whether the user can garner any useful information from them.

I would also add a lead in paragraph for this section that says a little about what "Global cache cleanup" means. It's not a very descriptive heading on its own.

pledbrook replied:
An entry for the global cache directory could also link directly to the section on how it's cleaned up.

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+[[directory_layout:gradle_user_home_directory:cache_cleanup]]
+=== Global cache cleanup
+
+Since Gradle 4.10, the Gradle user home directory is being cleaned up automatically. The cleanup is performed in the background when the Gradle daemon is stopped or shuts down. If using `--no-daemon`, it runs in the foreground after the build session with a visual progress indicator.
```
I would rephrase:

> From version 4.10 onwards, Gradle automatically/periodically cleans it's user home directory.

I would also hold back on using the passive so heavily. I'm not a "never use the passive" person, but its overuse does make prose less readable.

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+
+
+[[directory_layout:project_root_directory]]
+== Project root directory
```
I think mention that Gradle creates its own hidden directory in the root project of every build that it runs. That directory has the following structure.

eriwen replied:
+1. This means `buildSrc` too, not just project root.

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+<8> The project's <<build_lifecycle.adoc#sec:settings_file, settings file>>
+
+[[directory_layout:project_root_directory:cache_cleanup]]
+=== Project cache cleanup
```
Same comment about an awkward transition applies here too.  I would prefer fuller explanations for each of the directories. They don't have to be in-depth, but it seems odd explaining the directory structure without more information on what those directories do and how Gradle uses them.

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+[[directory_layout:project_root_directory:cache_cleanup]]
+=== Project cache cleanup
+
+Since Gradle 4.10, the project-specific cache directory is being cleaned up automatically. After building the project, version-specific cache directories in `.gradle/<gradle-version>/` are checked periodically (at most every 24 hours) for whether they are still in use. They are deleted if they haven't been used for 7 days.
```
Same suggested change as for the global cache cleanup.

pledbrook commented on subprojects/docs/src/docs/userguide/directory_layout.adoc:
```diff
@@ -0,0 +1,104 @@
+.
+├── caches // <1>
+│   ├── 4.8 // <2>
+│   ├── 4.9 // <2>
```
I don't think Asciidoctor automatically aligns annotation markers to the right. Could you please manually right-align them?

marcphilipp replied:
why should they be right aligned? I think that would make them harder to read.

eriwen replied:
:o: What I typically do here is try to align these callouts vertically in the left-most column. That said, I don't think we have any data to suggest this is better. 

I'm very happy that you use callouts at all. We finally can do this now that we're all asciidoc :)

marcphilipp commented:
I've added another commit that addresses most of the feedback. I've deliberately decided not to flesh out details in a few places so we don't expose implementation details that might change and should be irrelevant to our users.

@eriwen @pledbrook Please take another look! 🙂

pledbrook commented:
>  I've deliberately decided not to flesh out details in a few places so we don't expose implementation details that might change and should be irrelevant to our users.

If that's the case, why show the directory layout at all? And should we explicitly call out that the information provided is subject to change, if that's the case? I'm just a bit confused about this.

marcphilipp commented:
I created #6326 to collect improvement proposals. Please add items there.
